src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs

```diff
@@ -90,26 +94,72 @@ public void Abort()
                 AddWebSocketHeaders(request, secKeyAndSecWebSocketAccept.Key, options);
 
                 // Create the handler for this request and populate it with all of the options.
-                handler = new SocketsHttpHandler();
-                handler.Credentials = options.Credentials;
-                handler.Proxy = options.Proxy;
-                handler.CookieContainer = options.Cookies;
-                handler.SslOptions.RemoteCertificateValidationCallback = options.RemoteCertificateValidationCallback;
-                if (options._clientCertificates?.Count > 0) // use field to avoid lazily initializing the collection
+                // Try to use a shared handler rather than creating a new one just for this request, if
+                // the options are compatible.
+                if (options.Credentials == null &&
+                    !options.UseDefaultCredentials &&
+                    options.Proxy == null &&
+                    options.Cookies == null &&
+                    options.RemoteCertificateValidationCallback == null &&
+                    options._clientCertificates?.Count == 0)
                 {
-                    if (handler.SslOptions.ClientCertificates == null)
+                    disposeHandler = false;
+                    handler = s_defaultHandler;
+                    if (handler == null)
+                    {
+                        handler = new SocketsHttpHandler()
+                        {
+                            PooledConnectionLifetime = TimeSpan.Zero,
+                            UseProxy = false,
+                            UseCookies = false,
+                        };
+                        if (Interlocked.CompareExchange(ref s_defaultHandler, handler, null) != null)
```

Comment:
This is an interesting pattern.  Basically you're throwing away the newly created handler in case the static was already assigned.

Is this faster than just doing the standard double-locking pattern for singletons?

Reply:
It's optimistic vs pessimistic concurrency, with tradeoffs.  Both cases are identical once the field has been initialized, and in both cases every caller will end up using the same published object; where they differ is if there's a race to initialize.  In the double-checked locking / pessimistic case, you only run the initialization code once, but you a) need an object to lock on and b) end up blocking forward progress from all callers while the initialization happens.  In the lock-free / optimistic approach, you may end up with multiple threads running the initialization logic, but you don't need an object to lock on and you're guaranteed of forward progress, e.g. one thread getting swapped out while it holds the lock doesn't block any other threads from making forward progress.

Reply:
Ok.  Thx for the explanation. It's probably pretty cheap to do a `new SocketsHttpHandler()` anyways.